crypto/evp/p_lib.c

```diff
@@ -1400,6 +1400,48 @@ static int evp_pkey_reset_unlocked(EVP_PKEY *pk)
     return 1;
 }
 
+#ifndef FIPS_MODULE
+static int evp_pkey_reset_downgrade(EVP_PKEY *pk)
+{
+    if (pk == NULL)
+        return 0;
+
+    pk->type = EVP_PKEY_NONE;
+    pk->save_type = EVP_PKEY_NONE;
+
+    pk->ameth = NULL;
+    pk->engine = NULL;
+    pk->pmeth_engine = NULL;
+    pk->pkey.ptr = NULL;
+#  ifndef OPENSSL_NO_RSA
+    pk->pkey.rsa = NULL;
+#  endif
+#  ifndef OPENSSL_NO_DSA
+    pk->pkey.dsa = NULL;
+#  endif
+#  ifndef OPENSSL_NO_DH
+    pk->pkey.dh = NULL;
+#  endif
+#  ifndef OPENSSL_NO_EC
+    pk->pkey.ec = NULL;
+    pk->pkey.ecx = NULL;
+#  endif
+
+    pk->references = 1;
+    pk->save_parameters = 1;
+    pk->attributes = NULL;
+    memset(&pk->ex_data, 0, sizeof(CRYPTO_EX_DATA));
+    pk->keymgmt = NULL;
+    pk->keydata = NULL;
+    pk->dirty_cnt = 0;
+    memset(&pk->operation_cache, 0, sizeof(pk->operation_cache));
+    pk->dirty_cnt_copy = 0;
+    memset(&pk->cache, 0, sizeof(pk->cache));
```

Comment:
The only difference between this function and `evp_pkey_reset_unlocked` is that this one doesn't touch the lock, at all.  That's a bit over the top, I would rather see that this become a refactor of `evp_pkey_reset_unlocked`, and that the lock creation is moved back to `EVP_PKEY_new()`.

Reply:
Sounds good, I'll make these changes. Thanks!